### PR TITLE
Generate popup page for notification

### PR DESCRIPTION
**Pull request by AzamatG17**

### CODE REVIEW

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/App.xaml.cs [hunk 1]
```diff
@@ -25,7 +25,11 @@ public App()
 
             ConfigureServices();
 
+            Connectivity.ConnectivityChanged += Connectivity_ConnectivityChanged;
+
             MainPage = new AppShell();
+
+            //var signalRService = new SignalRService();
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/App.xaml.cs:32]: will this commentary be needed later?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+using System.Text;
+
+namespace CheckDrive.Mobile.Models
+{
+    public class ChatDto
+    {
+        public string HeadlineText { get; set; }
+        public string Title { get; set; }
+    }
+}
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs:12]: I think it would be better to put <;> after the namespace CheckDrive.Mobile.Models

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
+﻿using CheckDrive.Mobile.ViewModels;
+using Rg.Plugins.Popup.Pages;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:8]: an empty line is left
if there are unused ones, it would be good to run them too

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs [hunk 4]
```diff
@@ -15,6 +17,8 @@ public class SignalRService
 
         public SignalRService()
         {
+            ShowPopupAsync("Siz shu Gentra moshinani shu 15:00 vaqtta oldingizmi");
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:20]: vaqtta -> vaqtda

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
+﻿using CheckDrive.Mobile.Models;
+using System;
+using System.Collections.ObjectModel;
+using System.Threading.Tasks;
+using System.Windows.Input;
+using Xamarin.Forms;
+
+namespace CheckDrive.Mobile.ViewModels
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs:8]: here too  ;  It would be good to write down

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace CheckDrive.Mobile.Views
```
Comment (Boburm1rzo) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:12]: here too  ;  It would be good to write down

**Review by Boburm1rzo — CHANGES_REQUESTED** [review]

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs (hunk 6 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:12]: Xamarin forms doesn't support this technology

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs (hunk 5 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs:8]: Xamarin forms doesn't support this technology

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/App.xaml.cs (hunk 1 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/App.xaml.cs:32]: No, it is not needed. It is for testing Popup page, because Popup page will work runtime.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs (hunk 3 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs]: I fixed it

**Review by Boburm1rzo — APPROVED** [review]

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace CheckDrive.Mobile.Models
+{
+    public class ChatDto
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs:7]: rename it to MessageDto

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs [hunk 8]
```diff
@@ -80,7 +84,8 @@ private async Task ShowPopupAsync(string message)
                 await SecureStorage.SetAsync("popup_message", message);
                 await SecureStorage.SetAsync("popup_visible", "true");
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:85]: maybe it would be better to use local storage with some kind of SQLite? Rather than using Secure Storage? We could store all notifications there directly, Secure Storage is not designed for this I'm afraid. In Secure Storage we can hold token and other important information, but for larger ones maybe it's better to use InMemory storage such as SQLite

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml [hunk 9]
```diff
@@ -1,10 +1,11 @@
...
-       xmlns:views="clr-namespace:CheckDrive.Mobile.Views"
-       Title="CheckDrive.Mobile"
-       NavBarIsVisible="False">
+<Shell
+    x:Class="CheckDrive.Mobile.AppShell"
+    xmlns="http://xamarin.com/schemas/2014/forms"
+    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+    xmlns:views="clr-namespace:CheckDrive.Mobile.Views"
+    Title="CheckDrive.Mobile"
+    NavBarIsVisible="False">
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:8]: Update your formatter, we should all use the same formatting options, otherwise each change would be showing unnecessary changes in formatting which would be distracting the reviewer.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml [hunk 10]
```diff
@@ -24,9 +25,10 @@
             <Style Class="FlyoutItemLabelStyle" TargetType="Label">
                 <Setter Property="TextColor" Value="White" />
             </Style>
-            <Style ApplyToDerivedTypes="True"
-                   Class="FlyoutItemLabelStyle"
-                   TargetType="Layout">
+            <Style
+                ApplyToDerivedTypes="True"
+                Class="FlyoutItemLabelStyle"
+                TargetType="Layout">
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:31]: Update your formatter, we should all use the same formatting options, otherwise each change would be showing unnecessary changes in formatting which would be distracting the reviewer.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml [hunk 11]
```diff
@@ -49,9 +51,10 @@
             <!--
                 Custom Style you can apply to any Flyout Item
             -->
-            <Style ApplyToDerivedTypes="True"
-                   Class="MenuItemLayoutStyle"
-                   TargetType="Layout">
+            <Style
+                ApplyToDerivedTypes="True"
+                Class="MenuItemLayoutStyle"
+                TargetType="Layout">
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:57]: Update your formatter, we should all use the same formatting options, otherwise each change would be showing unnecessary changes in formatting which would be distracting the reviewer.

Comment (Mirazyzz) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:57]: style should be the same as it was before, attributes should start on the same level as the Element

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml [hunk 12]
```diff
@@ -72,17 +75,20 @@
...
+        <ShellContent
+            Title="Tarix"
+            ContentTemplate="{DataTemplate views:HistoryPage}"
+            Icon="icon_history.png"
+            Route="HistoryPage" />
+        <ShellContent
+            Title="Profil"
+            ContentTemplate="{DataTemplate views:ProfilePage}"
+            Icon="icon_user.png"
+            Route="ProfilePage" />
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:92]: The same formatting issue, attributes should start on the same level as the Element.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs (hunk 4 above)
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:20]: Where should we get this message from? If we want to merge this PR it should be production ready.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs [hunk 13]
```diff
@@ -80,7 +84,8 @@ private async Task ShowPopupAsync(string message)
...
 
-
+                var popup = new NotificationPopupPage(message);
+                await PopupNavigation.Instance.PushAsync(popup);
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:88]: can we use **INavigationService**?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+{
+    public class ChatViewModel : BaseViewModel
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs:10]: maybe **NotificationViewModel** would be a better name?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        public string Message
+        {
+            get => message;
+            set
+            {
+                message = value;
+                OnPropertyChanged();
+            }
+        }
+        public ICommand AcceptCommand { get; set; }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs:22]: separate with an empty line.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+        public async Task OnAccept()
+        {
+            Console.WriteLine("Accept button clicked");
+        }
+
+        public async Task OnCancel()
+        {
+            Console.WriteLine("Cancel button clicked");
+        }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs:41]: should those methods close popup?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/Errors/NoInternetErrorPage.xaml
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/Errors/NoInternetErrorPage.xaml:1]: please revert back formatting, I can't see what was changed and what was not changed.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/Errors/UnknownErrorPage.xaml
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/Errors/UnknownErrorPage.xaml:1]: please revert back formatting, I can't see what was changed and what was not changed.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml:1]: Good name for the popup, let's use this name in other places too, like PopupViewModel etc...

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+                    AcceptText="Ha"
+                    AcceptBackgroundColor="#1C375C"
+                    CancelCommand="{Binding CancelCommand}"
+                    CancelText="Yoq"
+                    CancelTextColor="Black"
+                    HeadlineText="Avtomobil qabul qilish"
+                    SupportingText="{Binding Message}"
+                    SupportingTextFontSize="18"
+                    BackgroundColor="White"
+                    />
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml:26]: Should we give red color we reject button?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+    public partial class NotificationPopupPage : PopupPage
+    {
+        private readonly ChatViewModel _chatViewModel;
+        public NotificationPopupPage(string massage)
+        {
+            InitializeComponent();
+
+            _chatViewModel = new ChatViewModel();
+            _chatViewModel.Message = massage;
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:17]: maybe it's better to pass message as a parameter rather than changing it separately?

File: CheckDrive.Mobile/Screenshot 2024-09-27 144838.png
Comment (Mirazyzz) [inline CheckDrive.Mobile/Screenshot 2024-09-27 144838.png:1]: remove screenshot and attach it to the PR description

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Please, provide necessary changes.

**Comment by Mirazyzz** [comment]
You upload screenshots as a comment to the PR

image

**Comment by Mirazyzz** [comment]
Also, merge latest master into your branch

**Comment by AzamatG17** [comment]
Screenshot popup page
Screenshot 2024-09-27 144838

**Comment by Mirazyzz** [comment]
> Screenshot popup page Screenshot 2024-09-27 144838

Let's make "Yo'q" a danger button and change background in **App.xaml** you can find the color to use for _Danger_

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs (hunk 7 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/ChatDto.cs:7]: corrected

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml (hunk 9 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:8]: but it works like this for me Format XAML, Source: [https://www.youtube.com/shorts/TCz2IDuNZE8](url)

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs (hunk 14 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs]: corrected

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs (hunk 15 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs]: corrected

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs (hunk 16 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ChatViewModel.cs]: corrected

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs (hunk 4 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:26]: This message will come from API.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs (hunk 8 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs]: I don't think there is any point in saving this data. If the user closed application without clicking on accept or rejected bottoms, all the same when he opens the application he will see the same popup page

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml (hunk 9 above)
Comment (Mirazyzz) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/AppShell.xaml:8]: You should configure this tool in Visual Studio to start attributes on the same line as the element

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs (hunk 4 above)
Comment (Mirazyzz) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs:20]: If this message comes from API then you should use mock service for now and take it from proper object, not hardcoded text.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/MessageDto.cs [hunk 19]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace CheckDrive.Mobile.Models
+{
+    public class MessageDto
+    {
+        public string HeadlineText { get; set; }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Models/MessageDto.cs:9]: Content/Description

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/NotificationViewModel.cs [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+                if (PopupNavigation.Instance.PopupStack.Any())
+                {
+                    await PopupNavigation.Instance.PopAsync(true);
+                    Console.WriteLine("Popup yopildi.");
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Popup yopishda xatolik: {ex.Message}");
+                throw new Exception($"Popup yopilmayapti... {ex.Message}");
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/NotificationViewModel.cs:68]: View model should not rethrow an exception, it should handle it. It should show an error page or error popup.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/NotificationViewModel.cs [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+            AcceptCommand = new Command(async () => await OnAccept());
+            CancelCommand = new Command(async () => await OnCancel());
+        }
+
+        public async Task OnAccept()
+        {
+            await ClosePopup();
+            await _navigationService.GoBackAsync();
+
+            Console.WriteLine("Accept button clicked");
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/NotificationViewModel.cs:44]: Do we need Console.WriteLine on production?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+
+            _chatViewModel = new NotificationViewModel(massage);
+            
+            BindingContext = _chatViewModel;
+        }
+
+        private async void PopupPage_BackgroundClicked(object sender, EventArgs e)
+        {
+            await Navigation.PopAsync();
+        }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:24]: Should we close when background clicked? What happens to the message if user doesn't accept it or reject?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs (hunk 13 above)
Comment (AzamatG17) [reply CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/SignalRService.cs]: corrected

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Please, provide necessary changes

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+using Rg.Plugins.Popup.Pages;
+using System;
+using Xamarin.Forms.Xaml;
+
+namespace CheckDrive.Mobile.Views
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+    public partial class NotificationPopupPage : PopupPage
+    {
+        private readonly NotificationViewModel _chatViewModel;
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/NotificationPopupPage.xaml.cs:11]: should be _notificationViewModel

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Fix NotificationService
